File: src/features/doctor/doctor.command.ts
import type { CliOptions } from '../../app/cli.models'
import { DoctorPresenter } from './doctor.presenter'
import { DoctorService } from './doctor.service'

export class DoctorCommand {
  constructor(
    private readonly service: DoctorService,
    private readonly presenter: DoctorPresenter
  ) {}

  async execute(options: CliOptions): Promise<number> {
    const report = await this.service.inspect(options)
    this.presenter.show(report)
    return report.failures === 0 ? 0 : 1
  }
}

File: lib/src/features/doctor/doctor_command.dart
import 'package:better_effect/better_effect.dart';

import '../../app/app_effect.dart';
import '../../app/cli_options.dart';
import 'doctor_presenter.dart';
import 'doctor_service.dart';

abstract interface class DoctorCommand {
  AppEffect<int> execute(CliOptions options);
}

final class DoctorCommandLive implements DoctorCommand {
  const DoctorCommandLive();

  @override
  AppEffect<int> execute(CliOptions options) => Effect.result((use) async {
    final report = await use.unwrap(use<DoctorService>().inspect(options));
    await use.unwrap(use<DoctorPresenter>().show(report));
    return report.failures == 0 ? 0 : 1;
  });
}
